README
LinuxRogue
README


Compiling LinuxRogue v0.1 on a modern Linux would throw up a zillion warnings
and errors. So, one weekend I sat down and cleaned up the code and this is the
result.

I have never played Rogue for more than a few levels so I don't know of any
errors that might come up in those levels. It shouldn't since I have not
changed anything in the logic. Please refer the CHANGES file for the changes
done.

All the best in your endeavour! :)


Ashwin N <ashwin@despammed.com>
---

This is an updated version of rogue 5.3-clone. It's a LINUX port which
has all of the posted patches added to it, and considerable numbers of
bug fixes. I've also added a help page and made it operate correctly
setuid.

Alan Cox [iiitac@pyr.swan.ac.uk]

Original readme entries...
----
I've ported this to LINUX..It seems to work fine.  I mostly
POSIXized machdep.c and got rid of some floating point 
arithmetic.  (Not a terribly difficult effort...)  This 
is a first effort at dealing with the POSIX terminal I/O
so it may not be perfect.  

I didn't add the patches to the original source code
so I don't have the bug fixes added.  I used the original
machdep.c instead of the new "SYSV" version so don't know
if my effort was totally in vain. Heck I got to learn something
about the new terminal I/O model so it wasn't all in vain ;-)

Steve Wilson stevew@netcom.com

----
Yes, folks, this is the same old rogue 5.3-clone source that I've posted
here before.  So why do it again?  Well, I got many complaints/inquisitions
about whether this code was ported to a number of different UNIX/ULTRIX/VMS
systems.  The answer is still "no."  BUTTTT..., I would like to first port
it to as may of the different UNIX systems as I can, and go from there.

My problem is that I'm not particularly familiar with the ins and outs of
other UNIX systems besides the one I work on. (BSD 4.[23])  So what I've
done is set up the rogue source code so that all of the machine dependent
stuff is localized in one source file, machdep.c.  Enclosed in machdep.c
are all of the routines that I thought were even slightly machine-dependent.
Also enclosed in machdep.c is a copious set of comments describing what's
going on in there.  Now, most of these routines I believe will run most
UNIX systems without modification, but then again, some won't.
That's where I need some help.  I need folks who have access to other
UNIX systems to make modifications to machdep.c that will effectively
port the entire rogue program to their particular UNIX system.  Enough
instructions are included in machdep.c to do this.  The modified machdep.c
can be e-mailed to me, whereby I can incorporate the changes, and post
versions of machdep.c to comp.sources.games as it becomes more complete.

Does this sound remotely interesting to you?  Please give me a hand.

If you can't do it yourself, maybe you know someone who can.  A little
thumbing through your system documentation is all you probably need.

I am hoping all the system-dependencies are in machdep.c, but I suspect
there may be some 'C' constructs that various compilers out there don't
support or treat differently.  I need to know about these too.  I'm not
expecting much of this on UNIX systems, though.

The code runs through 'lint' fine on my system.  All identifiers are unique
within their first eight characters.  Function names are unique within
their first seven characters.  All file names are under 14 characters.


Prerequisites of the operating system:

	32-bit long integer. (at least)
	16-bit short integer. (at least)

	Standard I/O library providing the following:
		stdio.h (include file),
		getchar(),
		fopen(),
		fclose(),
		fread(),
		fwrite(),
		fgets(),
		printf(),
		sprintf(),
		fflush(),
		rewind().

	String library providing the following:
		strcpy(),
		strncpy(),
		strcmp(),
		strncmp().

	The curses/termlib libraries are desireable, but not strictly necessary.
	The rogue source can be compiled with -DCURSES.  This will enable a
	self-contained curses/termlib emulation package.  This package has some
	restrictions which are outined in the comments near the top of the
	enclosed source file curses.c.  This curses emulation package was put
	into the source for systems that may not have curses available, or have
	very buggy versions.  If your system has usable working curses/termlib
	libraries, it is suggested that you use them instead of the enclosed
	curses library.


Tim Stoehr
tims@zeus.TEK.COM


CHANGES
LinuxRogue
CHANGES log


Version: 0.3.2
2004-05-02
Ashwin N. <ashwin@despammed.com>

Miguel G. C. <mgcuadrado@indra.es>
* Extended out the QWERTY specific key mapping into a generic configuration.
  Added support for DVORAK key mapping.

Ashwin N. <ashwin@despammed.com>
* Removed all (ancient) UNIX specific idiosyncracies in machdep.c

-------------------------------------------------------------------------------

Version: 0.3
2004-02-15
Ashwin N. <ashwin@despammed.com>

Ashwin N <ashwin@despammed.com>
* Header files added for all source files. #includes of these files added
  to all source files to replace the former dirty way of using externs.
* Makefile
  Prerequisites are generated automatically. Needs some more work.
* Corrected tons of punctuation mistakes in the messages displayed to the
  player.
* static on functions removed. It was unnecessary in the first place.
* Unused extern variable declarations removed.
* Lots of beautification of code (mostly on braces).
* shorts converted to ints in lots of places.
* spec_hit.c file renamed to special_hit.c

Chris
* hit.c
  Integer math bug fixed in damage to armor.
* inventory.c
  The output is always sorted by the selection letter. The other change allows
  you to select an item from the inventory viewing screen. The second is how
  the old inventory() function ended up moved to inv_sel() with inventory()
  just becoming a call to inv_sel() with arguments.
* monster.c
  Added a function fix_mon_damage. This function corrects the damage pointer
  when loading a saved game for a monster. Prior to this fix, monsters already
  present on a level when saved were restored with random damage amounts. Most
  commonly unable to hit the player due to their damage being 0.
* move.c
  My changes are an attempt to enable following of tunnels between rooms. It
  gets a little funny when you are in one of those maze style rooms but works
  for the most part. Following of turns in the tunnels is only active for
  control and a direction. Such as ^H.
* object.c
  The weapon equivalent of the monster fix. It was very annoying to restore a
  saved game and suddenly only be able to hit a monster with the damage for
  your strength.
* pack.c
  Contains the remainder of the changes to allow selection during inventory
  display.
* play.c
  The changes to allow the / command to identify characters on the screen as
  per rogue-guide.txt.
* save.c
  Added code to call the functions to fix monster and weapon damages when
  restoring a saved file.
* rogue-guide.txt
  High bit characters changed to normal chars.

-------------------------------------------------------------------------------

Version: 0.2.2
2004-01-15
Ashwin N <ashwin@despammed.com>

* V0.2.1 carried a corrupted AUTHORS file. The file has been repaired.

-------------------------------------------------------------------------------

Version: 0.2.1
2003-03-09
Ashwin N <ashwin@despammed.com>

Lots of fixes in 0.2.1. Most of the Segmentation Faults should disappear now.

* After seeking help (using ?) the dungeon map used to disappear. This is
  fixed.
* get_input_line was being called with the wrong parameters in lots of places.
  All of them are fixed. No more silly Segmentation Faults that were being
  reported should occur.
* Password has been changed from an eccentric one to "password"
* Command to create objects in wizard mode was Ctrl-C, now changed to Ctrl-/

  (Thanks to Matt <mattk@cyberonic.com> for pointing out all the above bugs
  and their fixes too.)

* Karl-Olov Serrander reported problems with getlogin() in function md_gln()
  under RH8. Replaced it with getpwnam() as he suggested.

-------------------------------------------------------------------------------

Version: 0.2
2003-02-21
Ashwin N <ashwin@despammed.com>

* The inbuilt curses/termlib emulation code (for systems which don't have
  ncurses) has been removed.
* All register variables changed to normal.
* All function declaration and definitions ANSI-fied from the old K&R style.
* All warnings fixed (a ton of them used to show up if I compiled with -Wall).

-------------------------------------------------------------------------------

Verison: 0.1

* The latest version I could get my hands on when I looked for a Linux port of
  Rogue.

-=-=-=-=-=-=-=-=-

Title:          Rogue Clone
Version:        LinuxRogue 0.1 (Rogue 5.3 Clone Pl2)
Description:    Rogue dungeon game
Author:         Alan Cox(anarchy@sunacm.swan.ac.uk)
                derived from prior work by
                        stevew@netcom.com(Steve Wilson)         (Linux Port)
                        tim@zeus.tek.com(Tim Stoehr)            (Original work)
Maintained-By:  Alan Cox(anarchy@sunacm.swan.ac.uk)
                (Swansea University Computer Society)
Platforms:      Linux 0.95 or better, Linux-Lite
Copying-Policy: Freely Redistributable
Keywords:       rogue dungeons of doom linux game yendor

-=-=-=-=-=-=-=-=-


src/score.c
/*
 * score.c
 *
 * This source herein may be modified and/or distributed by anybody who
 * so desires, with the following restrictions:
 *    1.)  No portion of this notice shall be removed.
 *    2.)  Credit shall not be taken for the creation of this source.
 *    3.)  This code is not to be traded, sold, or used for personal
 *         gain or profit.
 *
 */

#include <ncurses.h>
#include <stdio.h>
#include <string.h>
#include "rogue.h"
#include "hit.h"
#include "init.h"
#include "inventory.h"
#include "machdep.h"
#include "main.h"
#include "message.h"
#include "pack.h"
#include "ring.h"
#include "score.h"

char *score_file = ".roguescores";

extern char login_name[];
extern char *m_names[];
extern short max_level;
extern boolean score_only, show_skull, msg_cleared;
extern char *byebye_string, *nick_name;
extern fighter rogue;
extern struct id id_scrolls[];
extern struct id id_potions[];
extern struct id id_wands[];
extern struct id id_rings[];
extern struct id id_weapons[];
extern struct id id_armors[];

void center(short, char *);
void id_all(void);
void sell_pack(void);
void sf_error(void);
int name_cmp(char *, char *);
void insert_score(char [][82], char [][30], char *, int, int,
                  object *, int);
void nickize(char *, char *, char *);
int get_value(object *);

void killed_by(object *monster, int other)
{
	char buf[80];

	md_ignore_signals();

	if (other != QUIT)
	{
		rogue.gold = ((rogue.gold * 9) / 10);
	}

	if (other)
	{
		switch(other)
		{
		case HYPOTHERMIA:
			(void) strcpy(buf, "Died of hypothermia");
			break;
		case STARVATION:
			(void) strcpy(buf, "Died of starvation");
			break;
		case POISON_DART:
			(void) strcpy(buf, "Killed by a dart");
			break;
		case QUIT:
			(void) strcpy(buf, "Quit");
			break;
		}
	}
	else
	{
		(void) strcpy(buf, "Killed by ");
		if (is_vowel(m_names[monster->m_char - 'A'][0]))
		{
			(void) strcat(buf, "an ");
		}
		else
		{
			(void) strcat(buf, "a ");
		}
		(void) strcat(buf, m_names[monster->m_char - 'A']);
	}
	(void) strcat(buf, " with ");
	sprintf(buf+strlen(buf), "%ld gold", rogue.gold);
	if ((!other) && show_skull)
	{
		clear();
		mvaddstr(4, 32, "__---------__");
		mvaddstr(5, 30, "_~             ~_");
		mvaddstr(6, 29, "/                 \\");
		mvaddstr(7, 28, "~                   ~");
		mvaddstr(8, 27, "/                     \\");
		mvaddstr(9, 27, "|    XXXX     XXXX    |");
		mvaddstr(10, 27, "|    XXXX     XXXX    |");
		mvaddstr(11, 27, "|    XXX       XXX    |");
		mvaddstr(12, 28, "\\         @         /");
		mvaddstr(13, 29, "--\\     @@@     /--");
		mvaddstr(14, 30, "| |    @@@    | |");
		mvaddstr(15, 30, "| |           | |");
		mvaddstr(16, 30, "| vvVvvvvvvvVvv |");
		mvaddstr(17, 30, "|  ^^^^^^^^^^^  |");
		mvaddstr(18, 31, "\\_           _/");
		mvaddstr(19, 33, "~---------~");
		center(21, (nick_name[0] ? nick_name : login_name));
		center(22, buf);
	}
	else
	{
		message(buf, 0);
	}
	message("", 0);
	put_scores(monster, other);
}

void win(void)
{
	unwield(rogue.weapon);		/* disarm and relax */
	unwear(rogue.armor);
	un_put_on(rogue.left_ring);
	un_put_on(rogue.right_ring);

	clear();
	mvaddstr(10, 11, "@   @  @@@   @   @      @  @  @   @@@   @   @   @");
	mvaddstr(11, 11, " @ @  @   @  @   @      @  @  @  @   @  @@  @   @");
	mvaddstr(12, 11, "  @   @   @  @   @      @  @  @  @   @  @ @ @   @");
	mvaddstr(13, 11, "  @   @   @  @   @      @  @  @  @   @  @  @@    ");
	mvaddstr(14, 11, "  @    @@@    @@@        @@ @@    @@@   @   @   @");
	mvaddstr(17, 11, "Congratulations,  you have  been admitted  to  the");
	mvaddstr(18, 11, "Fighters' Guild.   You return home,  sell all your");
	mvaddstr(19, 11, "treasures at great profit and retire into comfort.");
	message("", 0);
	message("", 0);
	id_all();
	sell_pack();
	put_scores((object *) 0, WIN);
}

void quit(boolean from_intrpt)
{
	char buf[128];
	int i;
	int	orow = 0;
	int ocol = 0;
	boolean mc = 0;

	md_ignore_signals();

	if (from_intrpt)
	{
		orow = rogue.row;
		ocol = rogue.col;
		mc = msg_cleared;

		for (i = 0; i < DCOLS; i++)
		{
			buf[i] = mvinch(0, i);
		}
	}
	check_message();
	message("Really quit? (y/n)", 1);
	if (rgetchar() != 'y')
	{
		md_heed_signals();
		check_message();
		if (from_intrpt)
		{
			for (i = 0; i < DCOLS; i++)
			{
				mvaddch(0, i, buf[i]);
			}
			msg_cleared = mc;
			move(orow, ocol);
			refresh();
		}
		return;
	}
	if (from_intrpt)
	{
		clean_up(byebye_string);
	}
	check_message();
	killed_by((object *) 0, QUIT);
}

void put_scores(object *monster, int other)
{
	int i;
	int rank = 10;
	int found_player = -1;
	int n;
	int x;
	int ne = 0;
	char scores[10][82];
	char n_names[10][30];
	char buf[100];
	FILE *fp;
	long s;
	boolean failed = 0;
	char *mode = "r+w";

	turn_into_games();
	while ((fp = fopen(score_file, mode)) == NULL)
	{
		if (!failed)
		{
			mode = "w";
		}
		else
		{
			message("Cannot read/write/create score file.", 0);
			sf_error();
		}
		failed = 1;
	}
	turn_into_user();
	(void) xxx(1);

	for (i = 0; i < 10; i++)
	{
		if (((n = fread(scores[i], sizeof(char), 80, fp)) < 80) && (n != 0))
		{
			sf_error();
		} else if (n != 0) {
			xxxx(scores[i], 80);
			if ((n = fread(n_names[i], sizeof(char), 30, fp)) < 30)
			{
				sf_error();
			}
			xxxx(n_names[i], 30);
		} else {
			break;
		}
		ne++;
		if (!score_only)
		{
			if (!name_cmp(scores[i]+15, login_name))
			{
				x = 5;
				while (scores[i][x] == ' ')
				{
					x++;
				}
				s = lget_number(scores[i] + x);
				if (rogue.gold < s)
				{
					score_only = 1;
				}
				else
				{
					found_player = i;
				}
			}
		}
	}
	if (found_player != -1)
	{
		ne--;
		for (i = found_player; i < ne; i++)
		{
			(void) strcpy(scores[i], scores[i+1]);
			(void) strcpy(n_names[i], n_names[i+1]);
		}
	}
	if (!score_only)
	{
		for (i = 0; i < ne; i++)
		{
			x = 5;
			while (scores[i][x] == ' ')
			{
				x++;
			}
			s = lget_number(scores[i] + x);

			if (rogue.gold >= s)
			{
				rank = i;
				break;
			}
		}
		if (ne == 0)
		{
			rank = 0;
		}
		else
		{
			if ((ne < 10) && (rank == 10))
		   	{
				rank = ne;
			}
		}
		if (rank < 10)
		{
			insert_score(scores, n_names, nick_name, rank, (short) ne, monster,
			             other);
			if (ne < 10)
			{
				ne++;
			}
		}
		rewind(fp);
	}

	clear();
	mvaddstr(3, 30, "Top  Ten  Rogueists");
	mvaddstr(8, 0, "Rank   Score   Name");

	md_ignore_signals();

	(void) xxx(1);

	for (i = 0; i < ne; i++)
	{
		if (i == rank)
		{
			standout();
		}
		if (i == 9)
		{
			scores[i][0] = '1';
			scores[i][1] = '0';
		}
		else
		{
			scores[i][0] = ' ';
			scores[i][1] = i + '1';
		}
		nickize(buf, scores[i], n_names[i]);
		mvaddstr(i+10, 0, buf);
		if (rank < 10)
		{
			xxxx(scores[i], 80);
			fwrite(scores[i], sizeof(char), 80, fp);
			xxxx(n_names[i], 30);
			fwrite(n_names[i], sizeof(char), 30, fp);
		}
		if (i == rank)
		{
			standend();
		}
	}
	refresh();
	fclose(fp);
	message("", 0);
	clean_up("");
}

void insert_score(char scores[][82], char n_names[][30], char *n_name,
                  int rank, int n, object *monster, int other)
{
	int i;
	char buf[82];

	if (n > 0)
	{
		for (i = n; i > rank; i--)
		{
			if ((i < 10) && (i > 0))
			{
				(void) strcpy(scores[i], scores[i - 1]);
				(void) strcpy(n_names[i], n_names[i - 1]);
			}
		}
	}
	sprintf(buf, "%2d    %6d   %s: ", rank + 1, (int) rogue.gold, login_name); // (int) by me

	if (other)
	{
		switch(other)
		{
		case HYPOTHERMIA:
			(void) strcat(buf, "died of hypothermia");
			break;
		case STARVATION:
			(void) strcat(buf, "died of starvation");
			break;
		case POISON_DART:
			(void) strcat(buf, "killed by a dart");
			break;
		case QUIT:
			(void) strcat(buf, "quit");
			break;
		case WIN:
			(void) strcat(buf, "a total winner");
			break;
		}
	}
	else
	{
		(void) strcat(buf, "killed by ");
		if (is_vowel(m_names[monster->m_char - 'A'][0]))
		{
			(void) strcat(buf, "an ");
		}
		else
		{
			(void) strcat(buf, "a ");
		}
		(void) strcat(buf, m_names[monster->m_char - 'A']);
	}
	sprintf(buf + strlen(buf), " on level %d ",  max_level);
	if ((other != WIN) && has_amulet())
	{
		(void) strcat(buf, "with amulet");
	}
	for (i = strlen(buf); i < 79; i++)
	{
		buf[i] = ' ';
	}
	buf[79] = 0;
	(void) strcpy(scores[rank], buf);
	(void) strcpy(n_names[rank], n_name);
}

int is_vowel(char ch)
{
	return( (ch == 'a') || (ch == 'e') || (ch == 'i')
	        || (ch == 'o') || (ch == 'u') );
}

void sell_pack(void)
{
	object *obj;
	short row = 2, val;
	char buf[80];

	obj = rogue.pack.next_object;

	clear();
	mvaddstr(1, 0, "Value      Item");

	while (obj)
	{
		if (obj->what_is != FOOD)
		{
			obj->identified = 1;
			val = get_value(obj);
			rogue.gold += val;

			if (row < DROWS)
			{
				sprintf(buf, "%5d      ", val);
				get_desc(obj, buf+11);
				mvaddstr(row++, 0, buf);
			}
		}
		obj = obj->next_object;
	}
	refresh();
	if (rogue.gold > MAX_GOLD)
	{
		rogue.gold = MAX_GOLD;
	}
	message("", 0);
}

int get_value(object *obj)
{
	int wc;
	int val = 0;

	wc = obj->which_kind;

	switch(obj->what_is)
	{
	case WEAPON:
		val = id_weapons[wc].value;
		if ((wc == ARROW) || (wc == DAGGER) || (wc == SHURIKEN)
		    || (wc == DART))
		{
			val *= obj->quantity;
		}
		val += (obj->d_enchant * 85);
		val += (obj->hit_enchant * 85);
		break;
	case ARMOR:
		val = id_armors[wc].value;
		val += (obj->d_enchant * 75);
		if (obj->is_protected)
		{
			val += 200;
		}
		break;
	case WAND:
		val = id_wands[wc].value * (obj->oclass + 1);
		break;
	case SCROLL:
		val = id_scrolls[wc].value * obj->quantity;
		break;
	case POTION:
		val = id_potions[wc].value * obj->quantity;
		break;
	case AMULET:
		val = 5000;
		break;
	case RING:
		val = id_rings[wc].value * (obj->oclass + 1);
		break;
	}
	if (val <= 0)
	{
		val = 10;
	}
	return(val);
}

void id_all(void)
{
	int i;

	for (i = 0; i < SCROLLS; i++)
	{
		id_scrolls[i].id_status = IDENTIFIED;
	}
	for (i = 0; i < WEAPONS; i++)
	{
		id_weapons[i].id_status = IDENTIFIED;
	}
	for (i = 0; i < ARMORS; i++)
	{
		id_armors[i].id_status = IDENTIFIED;
	}
	for (i = 0; i < WANDS; i++)
	{
		id_wands[i].id_status = IDENTIFIED;
	}
	for (i = 0; i < POTIONS; i++)
	{
		id_potions[i].id_status = IDENTIFIED;
	}
}

int name_cmp(char *s1, char *s2)
{
	short i = 0;
	int r;

	while(s1[i] != ':')
	{
		i++;
	}
	s1[i] = 0;
	r = strcmp(s1, s2);
	s1[i] = ':';
	return(r);
}

void xxxx(char *buf, int n)
{
	int i;
	unsigned char c;

	for (i = 0; i < n; i++)
	{
		/* It does not matter if accuracy is lost during this assignment */
		c = (unsigned char) xxx(0);

		buf[i] ^= c;
	}
}

long xxx(boolean st)
{
	static long f, s;
	long r;

	if (st)
	{
		f = 37;
		s = 7;
		return(0L);
	}
	r = ((f * s) + 9337) % 8887;
	f = s;
	s = r;
	return(r);
}

void nickize(char *buf, char *score, char *n_name)
{
	int i = 15;
	int j;

	if (!n_name[0])
	{
		(void) strcpy(buf, score);
		return;
	}
	(void) strncpy(buf, score, 16);

	while (score[i] != ':')
	{
		i++;
	}

	(void) strcpy(buf + 15, n_name);
	j = strlen(buf);

	while (score[i])
	{
		buf[j++] = score[i++];
	}
	buf[j] = 0;
	buf[79] = 0;
}

void center(short row, char *buf)
{
	short margin;

	margin = ((DCOLS - strlen(buf)) / 2);
	mvaddstr(row, margin, buf);
}

void sf_error(void)
{
	message("", 1);
	clean_up("sorry, score file is out of order");
}
